Boxes implementation

dawoodalbadi opened the pull request:
Hello
This is follow up to pull #16

dawoodalbadi committed:
no compile error in boxes.cpp

dawoodalbadi committed:
no compile error in pdflib_core.cpp

It needs to be checked for its functionality if it is correct or not.

dawoodalbadi committed:
no compile errors in pslib.cpp but there are for sure some logical
errors need to be checked

dawoodalbadi committed:
No compile errors in scpageoutput

No compile errors in scpageoutput but for sure there are logical errors
such as shrink() which need to be re-implemented

dawoodalbadi committed:
No compile error in scribusview

dawoodalbadi committed:
No compile error in sctextstruct

dawoodalbadi committed:
Correction for all previous commit

There were some stupid mistak in previous commits and I modified it now

dawoodalbadi committed:
No compile error in svgexplugin

dawoodalbadi committed:
There is no compile error in whole project

we play around some compile error to make the code compiled first but it
has some logical problems.
The main logic error that we need to fix is in using linecontrol  struct

dawoodalbadi committed:
No compile error

No compile error in the code. The code was written by Vox is fixed from
syntax side but now it need to be complete.

dawoodalbadi committed:
Fixing travis continous integration

dawoodalbadi committed:
fix pull conflict

dawoodalbadi committed:
fixing continous integration

dawoodalbadi committed:
some modification

now the code can export text to pdf but it has problem in line break
more over I did not modify the code to deal with path text

vrhome committed:
Fix line width

now scribus export text to pdf with one bug only which is repeating some
chars after default break line

vrhome committed:
Implementation of render method

vrhome committed:
Render method (2 bug)

render method work nicely for screen with only 2 small bug. first one
the text get bolder when we ever we write more text. the other bug is
the text dosent stop at left edge of the frame as it suppose to do but
it goes outside the frame

vrhome committed:
bold issue in last commit

vrhome committed:
rendor method complete

dawoodalbadi commented:
current status for scribus 

![snapshot1](https://cloud.githubusercontent.com/assets/9442944/12032423/a7a9c220-ae31-11e5-929b-a8496f2d262c.png)

andreas-vox commented:
> Am 29.12.2015 um 10:41 schrieb dawoodalbadi notifications@github.com:
> 
> current status for scribus
> ...

Hi Dagwood!

Can you push your current status to your github repository? I tried to compile it yesterday but got still errors.

/Andreas

dawoodalbadi commented:
Hello Andreas
I push all my work to this pull request under new branch called box
Implementation
Last commit for me was render method complete
You can download my repository from this command
git clone https://github.com/dawoodalbadi/ScribusCTL.git

Then do checkout and choose boximplemintatin branch
This is the easy way to check out my code

Best Regards
DAWOOD
On Dec 29, 2015 3:26 PM, "andreas-vox" notifications@github.com wrote:

> > Am 29.12.2015 um 10:41 schrieb dawoodalbadi notifications@github.com:
> > 
> > current status for scribus
> > ...
> 
> Hi Dagwood!
> 
> Can you push your current status to your github repository? I tried to
> compile it yesterday but got still errors.
> 
> /Andreas
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/andreas-vox/ScribusCTL/pull/18#issuecomment-167768398
> .

vrhome committed:
Moving LineControl back to cpp file

dawoodalbadi commented:
I fix merge conflict and the code passed the compiles test

vrhome committed:
Add FIXME note

The problem of repeating some chars after break line

dawoodalbadi commented:
Andreas
I face two tricky problem that I couldn't solve

First one, you can see it in last commit that I push today (fixme). The
second one is about appearance of cursor position. I couldn't figure out
how I can implement method for cursor position appearance.

2015-12-29 18:26 GMT+04:00 Dawood al-badi dawood.albadi@gmail.com:

> Hello Andreas
> I push all my work to this pull request under new branch called box
> Implementation
> Last commit for me was render method complete
> You can download my repository from this command
> git clone https://github.com/dawoodalbadi/ScribusCTL.git
> 
> Then do checkout and choose boximplemintatin branch
> This is the easy way to check out my code
> 
> Best Regards
> DAWOOD
> On Dec 29, 2015 3:26 PM, "andreas-vox" notifications@github.com wrote:
> 
> > > Am 29.12.2015 um 10:41 schrieb dawoodalbadi notifications@github.com:
> > > 
> > > current status for scribus
> > > ...
> > 
> > Hi Dagwood!
> > 
> > Can you push your current status to your github repository? I tried to
> > compile it yesterday but got still errors.
> > 
> > /Andreas
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/andreas-vox/ScribusCTL/pull/18#issuecomment-167768398
> > .

andreas-vox commented:
Hi Dawood,

you can use TextLayout::boundingBox(pos) to get the bounding box of the current character. Depending on text direction, take the left  or right border. It uses the virtual method Box::boundingBox() to extract the rectangle from the box tree.
The opposite direction is handled by TextLayout::screenToPosition(), which uses Box::pointToPosition()

Best regards,
Andreas

> Am 30.12.2015 um 09:01 schrieb dawoodalbadi notifications@github.com:
> 
> Andreas
> I face two tricky problem that I couldn't solve
> 
> First one, you can see it in last commit that I push today (fixme). The
> second one is about appearance of cursor position. I couldn't figure out
> how I can implement method for cursor position appearance.
> 
> 2015-12-29 18:26 GMT+04:00 Dawood al-badi dawood.albadi@gmail.com:
> 
> > Hello Andreas
> > I push all my work to this pull request under new branch called box
> > Implementation
> > Last commit for me was render method complete
> > You can download my repository from this command
> > git clone https://github.com/dawoodalbadi/ScribusCTL.git
> > 
> > Then do checkout and choose boximplemintatin branch
> > This is the easy way to check out my code
> > 
> > Best Regards
> > DAWOOD
> > On Dec 29, 2015 3:26 PM, "andreas-vox" notifications@github.com wrote:
> > 
> > > > Am 29.12.2015 um 10:41 schrieb dawoodalbadi notifications@github.com:
> > > > 
> > > > current status for scribus
> > > > ...
> > > 
> > > Hi Dagwood!
> > > 
> > > Can you push your current status to your github repository? I tried to
> > > compile it yesterday but got still errors.
> > > 
> > > /Andreas
> > > 
> > > —
> > > Reply to this email directly or view it on GitHub
> > > https://github.com/andreas-vox/ScribusCTL/pull/18#issuecomment-167768398
> > > .
> > 
> > —
> > Reply to this email directly or view it on GitHub https://github.com/andreas-vox/ScribusCTL/pull/18#issuecomment-167954331.

vrhome committed:
Fix duplicate glyphs at end of line

dawoodalbadi commented:
hello 
current state for scribus as you can see 
![snapshot2](https://cloud.githubusercontent.com/assets/9442944/12061697/7c8f0926-afa6-11e5-932f-5852747bba25.png)
but still there are problem with cursor position which will be solved by my partner @IbrahimHumood 

I am starting to work on pdf but I am still using storyText::text beside boxes so now I need to eliminate any uses of text in pdf and make it depend completely in boxes 

![snapshot3](https://cloud.githubusercontent.com/assets/9442944/12061713/ec69521a-afa6-11e5-958f-a631f0811084.png)

dawoodalbadi commented:
Hello Andreas 

In pdf you said that it need major changes but what I can see it does not need much changes because our goal to make pdf production based on glyphs only so what we need to is to make pdflib_core receive the tree of glyphs and then draw it. No need for any checks in pdf like isbreak or istab and all this layout stuff should be in textFrame::layout. 
Please explain for me what you think if I am wrong.

Another problem i discover a new bug in my code with my solution for duplicate glyphs that when when I try to write first char is disappear and goes next line but in second char every thing goes right

Regards

IbrahimHumood commented:
Hello Mr.Vox,
I am Dawood's partner who suppose to work on cursor position.
Regard what did you said about using box::boundingBox to extract rectangle
from the box tree, I did two things :
1. After debugging the code I found that glyph::boundingBox always returns
   not expected results. In addition, the result in groupBox::boundingBox
   becomes invalid always. I deleted every thing in glyph::boundingBox and
   added this line:

return FRect(x(),y()+descent(), width()+1, height()+1);

However, it is still not valid. just I need you to see the modifications is
it good? because if yes then the problem from an other  place.
1. In addition to the above I noticed the ascent and descent returns zero
   always. So, I added initialization for ascender and descender in the method
   createGlyphBox in pageItem_textframe.h . I made the initlization as
   folowing:
   
   ```
   const CharStyle& s(run.style());
   result->setAscent(s.font().ascent(s.fontSize() / 10.0));
   result->setDescent(-s.font().descent(s.fontSize() / 10.0));
   ```

the values of the ascender and descender appears finally. But, the text now
appears above the text frame as in the screen shot.

do you have any recommendations?

Regards
​

On 5 January 2016 at 09:17, dawoodalbadi notifications@github.com wrote:

> Hello Andreas
> 
> In pdf you said that it need major changes but what I can see it does not
> need much changes because our goal to make pdf production based on glyphs
> only so what we need to is to make pdflib_core receive the tree of glyphs
> and then draw it. No need for any checks in pdf like isbreak or istab and
> all this layout stuff should be in textFrame::layout.
> Please explain for me what you think if I am wrong.
> 
> Another problem i discover a new bug in my code with my solution for
> duplicate glyphs that when when I try to write first char is disappear and
> goes next line but in second char every thing goes right
> 
> Regards
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/andreas-vox/ScribusCTL/pull/18#issuecomment-168901065
> .

vrhome committed:
No problem in first char

I am forced to remove one assert condition because it will not be true
for the first char ever

vrhome committed:
Bounding methode not complete

it draw cursor but always on the top of the frame so it calculate y
incorrectly

vrhome committed:
Cursor Works Fine

dawoodalbadi commented:
Cursor work fine but did not respond to mouse click

Now I will try to fix that. 
God help me..!

Regards 
Dawood

vrhome committed:
Enhancement in Boundingbox

dawoodalbadi commented:
current state in scribus

https://drive.google.com/file/d/0B2zhep1TjhN_NUJ1Qkx1QktUU0k/view?usp=sharing

vrhome committed:
Fix pointToPosition()

vrhome committed:
Cleaning Boxes for unnecessary codes

dawoodalbadi commented:
point to position is solved now scribus detecting mouse click and move the cursor accordingly but still there are one bug which is wrong behave when you click at the end of line.
regards
dawood

vrhome committed:
Select text works for first line only

luzpaz commented:
:+1: 
@dawoodalbadi :smiley:

vrhome committed:
select text is now working correct

I notice we still have problem in mouse click that will cause sometimes
wrong behaveior in selection.

vrhome committed:
Modify render method

now all setting happen in glyphbox render method

vrhome committed:
Remove unused variable

vrhome committed:
Fix access of free'd memoery

vrhome committed:
Tab not working

dawoodalbadi commented:
Hello @andreas-vox 
I am working on Tab drawing but unfortunately I couldn't understand why it always sets layout.xadvance = 0 for the tab and for that it always draw the tab with zero width.If you have any idea how I can fix this please tell me. you can see my last modification for the tab in last commit. 
https://github.com/dawoodalbadi/ScribusCTL/commit/9011ee19a981b785b4535b9f33f1854ad658de18

BEST REGARDS
DAWOOD

vrhome committed:
Tab is work correctly

dawoodalbadi commented:
Drawing Tab problems has been solved 
Ignore last comment 

Regards
Dawood

vrhome committed:
Some cleanup

vrhome committed:
Fix verticalAlignment

vrhome committed:
Fix assumption of making glyphBox representing one char

I put the ground to fix this problem but I did not modify all codes
accordingly because I did not Know all the cases that we make this
assumption. What I did is that I add firstChar and lastChar in glyphrun
that will indicate how many chars in this run and for now I set
firstChar = lastChar = index of Char in story text. I need to modify any
function that send index of the char to pageitem to make it send also
first and last char index of the run if we have more than one char for
that run.

vrhome committed:
ToDo

IbrahimHumood committed:
Trying to fix justification

vrhome committed:
Starting to implement general painter.

I am trying to implement specific painter for each class that need to
render boxes for example screen painter and pdf painter and all these
clases will inherant from a general painter called ScPainter

vrhome committed:
Text alignment work correctly

dawoodalbadi commented:
Hello 
Last commit is fixing text alignment
In my tests I found that it remains only 2 problems in text layout which are paragraph effects and some fonts causes crashes. 

Best regards 
Dawood

Fahad-Alsaidi commented:
Great @dawoodalbadi & @IbrahimHumood . I notice also that when you select a big text, the selection is very slow and it goes char by char.

vrhome committed:
Fixing DropCap and some cleaning(not complete)

I couldent fix the offset correctly so I leave it for anyone can

dawoodalbadi commented:
Last commit is for fixing drop cap problem but there are problem in the offset I couldn't find out why it happen so I hope @andreas-vox will solve it

vrhome committed:
Fix Paragraph effects bullet mode

Fahad-Alsaidi commented:
I found another bug, when there is more than one columns in the text frame, there are a big gap between the columns. 
How to test:
- create a text frame.
- fill it with contents.
- in columns and text distances section, put columns to 2 or 3 or 4 and notice.

I found this bug by testing this [file](http://scribusstuff.org/content/show.php/Flyer+for+events+-+tri-fold?content=158454)

vrhome committed:
Fix multicolumns

vrhome committed:
No need for reference here

dawoodalbadi commented:
problem which reported by @Fahad-Alsaidi  has been solved.
I welcome any other reports 

regards
Dawood Albadi

vrhome committed:
Solve problem of ascent and decent in some fonts

luzpaz commented:
I will try to compile and test as well.

dawoodalbadi commented:
Thank you @luzpaz 
Mr @Fahad-Alsaidi has already reported 4 new bugs 
1- writing lines of numbers will act very strange behavior
2- fixed nano font have problems in drawing selection
3- export to pdf will terminate scribus
4- in multi column test the cursor does not move the any columns other than the first

I welcome any new reports 

regards 
dawood

vrhome committed:
Export to pdf work but not perfect

pdf export work correctly but there still some feature doesnt work in
pdf but I dont want to solve that now because I am planing to implement
new render for the pdf and new class called pdfpainter and all feature
will be solved there.

dawoodalbadi commented:
Export to pdf has been solved but there are some problems in layout which will be solved when I implement pdf render method later.

regards

vrhome committed:
Fix cursor wrong behavior when write lines of numbers

dawoodalbadi commented:
problem 1,2,3 has been solved.  the behavior now similar to svn version
Mr. @Fahad-Alsaidi  you are welcome for further testing

Fahad-Alsaidi commented:
Thanks @dawoodalbadi for the update. However, I found another bug: Coloring text. Right now I cant nor change text color in one text frame though the color is exported to pdf ( only one color).

To summarise the bugs I face:
1. cursor movement is very buggy.
2. the selection is very difficult specially the last char 
3. no color text in one text frame.
4. pdf export doesn't export all colors.
5. when the text frame background is colored, it export strangely to pdf.

vrhome committed:
Fix setting text color

vrhome committed:
Fix move cursor to the next column

dawoodalbadi commented:
Hello @Fahad-Alsaidi 
Regarding bug 1 and 2, they are same problem caused by inaccurate position calculated for the click along with inaccurate bounding box for the glyphs but that is good enough for this stage. About text color and cursor position for multi columns text, they have been solved. About pdf, there are no layout setting at all in pdf now and all of that will be implemented in pdf_render method later. For now I am working to solve any bugs regarding text frame layout and after that I will move to pdf

Best regards

Fahad-Alsaidi commented:
Great ... I will leave pdf testing after you finish you work on it. 

Right now, I've discovered some functionalities don't work under colors and effects section which are:
1. underline text.
2. underline words only.
3. small caps.
4. strike out.
5. shadowed text
6. Right to left writing.

I will let you know if get any missing feature & please test all functionalities carefully.

Fahad-Alsaidi commented:
another problem, there is space above the text in the text frame, please see this screen shot.
![scribus-boxes-bugs2](https://cloud.githubusercontent.com/assets/1632566/12813771/b098ccca-cb53-11e5-94bf-3d84130b4372.jpg)

Fahad-Alsaidi commented:
@luzpaz have you succeed to build this pull request? Have you find any missing functionalities?

luzpaz commented:
@Fahad-Alsaidi & @dawoodalbadi 
Tried to build on OSX 10.10.5 

```
[  2%] Building CXX object scribus/third_party/lib2geom/CMakeFiles/scribus_2geom_lib.dir/svg-path.cpp.o
/tmp/scribusctltextbox20160204-17370-1lfvsko/scribus/text/boxes.cpp:102:22: error: use of undeclared identifier 'm_gylphs'
        for (int i = 0; i < m_gylphs.length(); ++i)
                            ^
/tmp/scribusctltextbox20160204-17370-1lfvsko/scribus/text/boxes.cpp:104:17: error: use of undeclared identifier 'm_glyphs'
                qreal width = m_glyphs[i].xadvance;
                              ^
/tmp/scribusctltextbox20160204-17370-1lfvsko/scribus/text/boxes.cpp:117:15: error: reference to non-static member function must be called; did you mean to call it with no arguments?
        int relPos = firstChar - pos;
                     ^~~~~~~~~
                              ()
/tmp/scribusctltextbox20160204-17370-1lfvsko/scribus/text/boxes.cpp:121:12: error: use of undeclared identifier 'm_glyphs'
                xPos1 += m_glyphs[i].xadvance;
                         ^
/tmp/scribusctltextbox20160204-17370-1lfvsko/scribus/text/boxes.cpp:123:16: error: use of undeclared identifier 'm_glyphs'
        qreal width = m_glyphs[relPos].xadvance;
                      ^
/tmp/scribusctltextbox20160204-17370-1lfvsko/scribus/text/boxes.cpp:127:12: error: use of undeclared identifier 'm_glyphs'
                xPos2 += m_glyphs[i].xadvance;
                         ^
/tmp/scribusctltextbox20160204-17370-1lfvsko/scribus/text/boxes.cpp:129:23: error: reference to non-static member function must be called; did you mean to call it with no arguments?
        return FRect(xPos1, -descent, xPos2 - xPos1, ascent);
                             ^~~~~~~
                                    ()
/tmp/scribusctltextbox20160204-17370-1lfvsko/scribus/text/boxes.cpp:129:47: error: reference to non-static member function must be called; did you mean to call it with no arguments?
        return FRect(xPos1, -descent, xPos2 - xPos1, ascent);
                                                     ^~~~~~
                                                           ()
8 errors generated.
make[2]: *** [scribus/text/CMakeFiles/scribus_text_lib.dir/boxes.cpp.o] Error 1
make[1]: *** [scribus/text/CMakeFiles/scribus_text_lib.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

dawoodalbadi commented:
hello @luzpaz 
I think you are trying to build master branch but my work is in boxesImplementation 
so make sure that you checkout right branch

best regards

luzpaz commented:
@dawoodalbadi I got it to compile. You were right I wasn't compiling your branch :/
I want to confirm @Fahad-Alsaidi observations also don't work for me. 

```
1. underline text.
2. underline words only.
3. small caps.
4. strike out.
5. shadowed text
6. Right to left writing.
```

luzpaz commented:
FYI, I spoke with MrB and he mentioned that 1.5.1svn will be released within a week or so. So rebasing this branch with master would be really good timing to coincide with that. It will also mean that more folks will test it (possibly including the devs).

dawoodalbadi commented:
Hello @luzpaz
I will do that this week because it seems that Andreas is to buzy and I have a limited time to finish this task 

Regards
Dawood

vrhome committed:
Fix underline problems

vrhome committed:
Fix strike out problem

vrhome committed:
Fix shadow text problem

vrhome committed:
Fix small caps problems

dawoodalbadi commented:
Hello guys

I solved issues from 1 to 5 and I leave last one because we will not need this feature in ctl. If you dont have any bug reports regarding text frame layout I will move to do some inhancment in the code and then I will move to pdf render implementation.

Fahad-Alsaidi commented:
Great @dawoodalbadi  & @IbrahimHumood we are getting closer. 
Here are my last bugs that I could find in text frame:
1.  there is still space above the text. This space makes old documents needs reformat from beginning. check and Full Mag issues and compare by yourself.
2. in paragraph effects: numerated list doesn't work, please also notice selection problem there.

dawoodalbadi commented:
Hello @Fahad-Alsaidi 

In case 1: I notice the problem and that happen only for some fonts but in general most fonts behave as svn version but I will look at this problem later.
case 2: all paragraph effects have problems especially when you try to undo the effect and this problem in my argent agenda for these days 
I have also other bugs regarding selection and cursor drawing need to fixed as soon as possible 

I welcome further testing in text frame layout 

regard 
dawood

vrhome committed:
Select drobcaps better than befor but stil not berfect

vrhome committed:
Solve crash problem in undo of bullet paragraph style

dawoodalbadi commented:
Last commit was to fix a crash regarding remove paragraph styles bullet or numbers but still there are one problem in number style which is repeating the number twice 

Best Regards

vrhome committed:
fix dublicate in paragraph style number bullet
